org-roam-extra.el

```diff
@@ -0,0 +1,268 @@
+  ;; Handle org-roam-capture part
+  (let* ((kwds (if (listp org-roam-extra-preformat-keywords) ; normalize org-roam-extra-preformat-keywords
+                   org-roam-extra-preformat-keywords
+                 (list org-roam-extra-preformat-keywords)))
+         ;; org-capture
+         (tp (nth 4 template))          ; org-capture template string
+         (plst (cdr template))         ; org-roam capture properties are here
+         (rx "\\(%\\^{[[:alnum:]-_]*}\\)") ; regexp for org-capture prompt wildcard
+         lst)
+    ;; First run:
+    ;; 1) Make a list of (rplc-s field-value match-position) for the second run
+    ;; 2) replace org-roam-capture wildcards
+    (dolist (kwd kwds)
+      (let* ((keyword (or (car-safe kwd) kwd))        ; prompt wildcard keyword
+             (field-name (or (cdr-safe kwd) kwd)) ; bibtex field name
+             (field-value                ; get the bibtex field value
+              (or (s-format (concat "${" field-name "}") 'bibtex-completion-apa-get-value entry)
+                  nil))                                         ; nil will be used to set back the proper wildcard
+             (rplc-s (concat "%^{" (or keyword "citekey") "}")) ; org-capture prompt wildcard
+             (rplc-s2 (concat "${" (or keyword "citekey") "}")) ; org-roam-capture prompt wildcard
+             (head (plist-get plst :head))  ; org-roam-capture :head template
+             (fl-nm (plist-get plst :file-name)) ; org-roam-capture :file-name template
+             (i 1)                               ; match counter
+             pos)
+        ;; Search for rplc-s, set flag m if found
+        (while (string-match rx tp pos)
+          (if (string= (match-string 1 tp) rplc-s)
+              (progn
+                (setq pos (length tp))
+                (pushnew (list rplc-s field-value i) lst ))
+            (setq pos (match-end 1)
+                  i (1+ i))))
+        ;; Replace org-roam-capture prompt wildcards
+        (when (and field-value head )
+          (plist-put plst :head (s-replace rplc-s2 field-value head)))
+        (when (and field-value fl-nm)
+          (plist-put plst :file-name (s-replace rplc-s2 field-value fl-nm)))))
+    ;; Second run: replace prompts and propmt matches in org-capture template string
+    (dolist (l lst)
+      (when (and (nth 1 l) (stringp tp))
+        (let ((pos (concat "%\\" (number-to-string (nth 2 l)))))
+          ;; replace prompt match wildcards with prompt wildcards
+          ;; replace prompt wildcards with bitex field value
+          (setq tp (s-replace pos (car l) tp)
+                tp (s-replace (car l) (nth 1 l) tp))))
+      (setf (nth 4 template) tp))
+    template))
+
+(defun org-roam-extra--switch-perspective ()
+  "Helper function for `org-roam-extra-org-ref-note-fn'."
+  (when
+      (and (featurep 'projectile)
+           (require 'projectile nil t)
+           (featurep 'persp-mode)
+           (require 'persp-mode nil t))
```

Comment:
not sure if the `featurep` checks are needed, I think just the `require` clauses are needed.